### PR TITLE
Port waitForPendingWrites from android

### PR DESCRIPTION
**Pull request by wu-hui**
* Port waitForPendingWrites

### CODE REVIEW

File: packages/firestore/src/api/database.ts [hunk 1]
```diff
@@ -478,6 +478,24 @@ export class Firestore implements firestore.FirebaseFirestore, FirebaseService {
     return this._firestoreClient!.clientShutdown;
   }
 
+  /**
+   * Waits until all currently pending writes for the active user have been acknowledged by the
+   * backend.
+   *
+   * The returned Promise completes immediately if there are no outstanding writes. Otherwise, the
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts]: s/completes/resolves/

File: packages/firestore/src/api/database.ts [hunk 2]
```diff
@@ -478,6 +478,24 @@ export class Firestore implements firestore.FirebaseFirestore, FirebaseService {
...
+   * session), but it does not wait for writes that were added after the method is called. If you
+   * wish to wait for additional writes, you have to call `waitForPendingWrites()` again.
+   *
+   * Any outstanding `waitForPendingWrites()` Promises are rejected during user changes.
+   *
+   * @return A `Promise` which resolves when all currently pending writes have been
+   * acknowledged by the backend.
+   */
+  _waitForPendingWrites(): Promise<void> {
+    return this._firestoreClient!.waitForPendingWrites();
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts]: Please call `ensureClientConfigured()` to make sure this works even when it is the first operation that is scheduled.

File: packages/firestore/src/core/firestore_client.ts [hunk 3]
```diff
@@ -522,6 +522,20 @@ export class FirestoreClient {
     });
   }
 
+  /**
+   * Returns a `Promise` resolves when all the pending writes at the time when this method is called
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts]: I would remove the backticks for Promise. It's just a normal term in JS, not a class or method name (also it should be "Promise that resolves");

Also: s/when all the pending writes at the time/when all writes that were pending at the time/

File: packages/firestore/src/core/sync_engine.ts [hunk 4]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     }
   }
 
+  /**
+   * Takes a snapshot of current mutation queue, and register a user callback
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: s/register/registers

File: packages/firestore/src/core/sync_engine.ts [hunk 5]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+   * which will be called when all those mutations in the snapshot are either
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: What do you mean by "snapshot" here? Snapshot usually means something different in our code base.

File: packages/firestore/src/core/sync_engine.ts [hunk 6]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+  async registerPendingWritesCallback(callback: Deferred<void>): Promise<void> {
+    if (!this.remoteStore.canUseNetwork()) {
+      log.debug(
+        LOG_TAG,
+        'The network is disabled. The task returned by ' +
+          "'awaitPendingWrites()' will not complete until the network is enabled."
+      );
+    }
+
+    const largestBatchId = await this.localStore.getHighestUnacknowledgedBatchId();
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: Nit: Make this `highestBatchId`.

File: packages/firestore/src/core/sync_engine.ts [hunk 7]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     }
   }
 
+  /**
+   * Takes a snapshot of current mutation queue, and register a user callback
+   * which will be called when all those mutations in the snapshot are either
+   * accepted or rejected by the server.
+   */
+  async registerPendingWritesCallback(callback: Deferred<void>): Promise<void> {
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: This method doesn't seem to have to be async. You can just return void.

File: packages/firestore/src/core/sync_engine.ts [hunk 8]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+
+    const largestBatchId = await this.localStore.getHighestUnacknowledgedBatchId();
+
+    if (largestBatchId === BATCHID_UNKNOWN) {
+      // Trigger the callback right away if there is no pending writes at the moment.
+      callback.resolve();
+      return Promise.resolve();
+    }
+
+    const callbacks = this.pendingWritesCallbacks.get(largestBatchId) || [];
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: It would be beneficial to verify what the transpiled JS looks like, but this could potentially be:

```
const callbacks = [... this.pendingWritesCallbacks.get(largestBatchId), callback]
```

If we don't transpile the spread operator away, there is some chance we are breaking users, so this might be not worth exploring.

File: packages/firestore/src/core/sync_engine.ts [hunk 9]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+    callbacks.push(callback);
+    this.pendingWritesCallbacks.set(largestBatchId, callbacks);
+    return Promise.resolve();
+  }
+
+  /**
+   * Triggers callbacks waiting for this batch id to get acknowledged by server,
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: s/callbacks/the callbacks that are/

File: packages/firestore/src/core/sync_engine.ts [hunk 10]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+   * if there are any.
+   */
+  private triggerPendingWritesCallbacks(batchId: BatchId): void {
+    (this.pendingWritesCallbacks.get(batchId) || []).forEach(callback => {
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: If you use a `for ... of` loop then you don't need to create an empty array here.

File: packages/firestore/src/core/sync_engine.ts [hunk 11]
```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
...
+  }
+
+  /** Reject all outstanding callbacks waiting for pending writes to complete. */
+  private rejectOutstandingPendingWritesCallbacks(): void {
+    this.pendingWritesCallbacks.forEach(callbacks => {
+      callbacks.forEach(callback => {
+        callback.reject(
+          new FirestoreError(
+            Code.CANCELLED,
+            "'waitForPendingWrites' task is cancelled due to User change."
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: This might be applicable to Android to: Since this is a very generalized function it might make sense to pass in the error (at least the message if you want to keep the stacktrace).

Also nit: s/User/a user/.

File: packages/firestore/src/local/indexeddb_mutation_queue.ts [hunk 12]
```diff
@@ -265,6 +265,26 @@ export class IndexedDbMutationQueue implements MutationQueue {
       .next(() => foundBatch);
   }
 
+  getHighestUnacknowledgedBatchId(
+    transaction: PersistenceTransaction
+  ): PersistencePromise<BatchId> {
+    const range = IDBKeyRange.bound(
+      [this.userId, Number.NEGATIVE_INFINITY],
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/indexeddb_mutation_queue.ts]: While this is fine, the method below uses `BATCHID_UNKNOWN` which also works.

File: packages/firestore/src/local/indexeddb_mutation_queue.ts [hunk 13]
```diff
@@ -265,6 +265,26 @@ export class IndexedDbMutationQueue implements MutationQueue {
       .next(() => foundBatch);
   }
 
+  getHighestUnacknowledgedBatchId(
+    transaction: PersistenceTransaction
+  ): PersistencePromise<BatchId> {
+    const range = IDBKeyRange.bound(
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/indexeddb_mutation_queue.ts]: Is it possible to just use `upperBound`?

File: packages/firestore/test/integration/api/database.test.ts [hunk 14]
```diff
@@ -1098,6 +1101,21 @@ apiDescribe('Database', (persistence: boolean) => {
     });
   });
 
+  it('can unlisten queries after shutdown', async () => {
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: Please move these changes to a different PR. We don't want to revert two PRs worth of work if we happen to find a problem with one of them.

File: packages/firestore/test/integration/api/database.test.ts [hunk 15]
```diff
@@ -1120,4 +1138,50 @@ apiDescribe('Database', (persistence: boolean) => {
       }).to.throw();
     });
   });
+
+  it('can wait for pending writes as expected', async () => {
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: "as expected" should be redundant for our tests :)

File: packages/firestore/test/integration/api/database.test.ts [hunk 16]
```diff
@@ -1120,4 +1138,50 @@ apiDescribe('Database', (persistence: boolean) => {
...
+      // Prevent pending writes receiving acknowledgement.
+      await firestore.disableNetwork();
+
+      const awaitPendingWrites1 = waitForPendingWrites(firestore);
+      const pendingWrites = docRef.set({ foo: 'bar' });
+      const awaitPendingWrites2 = waitForPendingWrites(firestore);
+
+      // `awaitsPendingWrites1` resolves immediately because there are no pending writes at
+      // the time it is created.
+      await expect(awaitPendingWrites1).to.be.eventually.fulfilled;
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: Hm. I don't know how you can check in JS if a Promise is already resolved. If you want to keep this line (I think you should) I would adjust the comment to explain why the Promise does resolve even though the network is offline.

File: packages/firestore/test/integration/api/database.test.ts [hunk 17]
```diff
@@ -1120,4 +1138,50 @@ apiDescribe('Database', (persistence: boolean) => {
...
+
+      // pending writes can receive acknowledgements now.
+      await firestore.enableNetwork();
+      await expect(pendingWrites).to.be.eventually.fulfilled;
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: `await expect(pendingWrites).to.be.eventually.fulfilled` is the same as `await pendingWrites`.

File: packages/firestore/test/integration/api/database.test.ts [hunk 18]
```diff
@@ -1120,4 +1138,50 @@ apiDescribe('Database', (persistence: boolean) => {
...
+  it('waiting for pending writes should fail when user changes', async () => {
+    await withMockCredentialProviderTestDb(persistence, async (db, mockCredentialsProvider) => {
+      // Prevent pending writes receiving acknowledgement.
+      await db.disableNetwork();
+      db.doc('abc/123').set({ foo: 'bar' });
+      const awaitPendingWrite = waitForPendingWrites(db);
+
+      mockCredentialsProvider.changeUserTo(new User('user_1'));
+
+      await expect(awaitPendingWrite).to.be.eventually.rejected;
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: Please do test the error message as well. It will also make the test much clearer.

File: packages/firestore/test/integration/api/database.test.ts [hunk 19]
```diff
@@ -1120,4 +1138,50 @@ apiDescribe('Database', (persistence: boolean) => {
...
+
+  it('waiting for pending writes resolves immediately when offline and no pending writes',
+    async () => {
+      await withTestDoc(persistence, async docRef => {
+        const firestore = docRef.firestore;
+        // Prevent pending writes receiving acknowledgement.
+        await firestore.disableNetwork();
+
+        const awaitPendingWrites = waitForPendingWrites(firestore);
+        await expect(awaitPendingWrites).to.be.eventually.fulfilled;
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: This is the same test case as above. In general, it is usually better to have smaller and more focused test cases. Maybe we should remove the "offline but no writes" case from above and only test it here?

File: packages/firestore/test/integration/util/helpers.ts [hunk 20]
```diff
@@ -124,6 +126,20 @@ export const apiDescribe = apiDescribeInternal.bind(
 apiDescribe.skip = apiDescribeInternal.bind(null, describe.skip);
 apiDescribe.only = apiDescribeInternal.bind(null, describe.only);
 
+export class MockCredentialsProvider extends EmptyCredentialsProvider {
+
+  private listener: CredentialChangeListener | null = null;
+
+  changeUserTo(user: User): void {
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/util/helpers.ts]: Do you think it makes sense to call this `handleUserChange` to match SyncEngine?

File: packages/firestore/test/integration/util/helpers.ts [hunk 21]
```diff
@@ -163,6 +179,20 @@ export function withTestDb(
   });
 }
 
+export function withMockCredentialProviderTestDb(
+  persistence: boolean,
+  fn: (db: firestore.FirebaseFirestore, mockCredential: MockCredentialsProvider) => Promise<void>
+): Promise<void> {
+  const mockCredentialsProvider = new MockCredentialsProvider();
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const settings: any = { ...DEFAULT_SETTINGS };
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/util/helpers.ts]: You can potentially remove the `any`:

```
const settings = {... DEFAULT_SETTINGS, credentials: { client: mockCredentialsProvider, type: 'provider' }}
```

File: packages/firestore/test/integration/api/database.test.ts (hunk 17 above)
Comment (wu-hui) [reply packages/firestore/test/integration/api/database.test.ts]: Done.

File: packages/firestore/test/integration/api/database.test.ts (hunk 19 above)
Comment (wu-hui) [reply packages/firestore/test/integration/api/database.test.ts]: Done.

File: packages/firestore/test/integration/api/database.test.ts (hunk 18 above)
Comment (wu-hui) [reply packages/firestore/test/integration/api/database.test.ts]: Done.

File: packages/firestore/test/integration/util/helpers.ts (hunk 20 above)
Comment (wu-hui) [reply packages/firestore/test/integration/util/helpers.ts]: `handleUserChange` implies it's handling something, which is not true. Changed to `triggerUserChange` instead.

File: packages/firestore/test/integration/util/helpers.ts (hunk 21 above)
Comment (wu-hui) [reply packages/firestore/test/integration/util/helpers.ts]: Done.

File: packages/firestore/src/core/sync_engine.ts (hunk 7 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/core/sync_engine.ts]: Please follow the pattern that we use here: https://github.com/firebase/firebase-js-sdk/pull/2063/files#diff-3bab365c04d7fc7ae24002035afdd01fR614

It makes the interactions in the client simpler if we limit async methods to their smallest possible scope. Right now, registerPendingWritesCallback only has one consumer, but once it has two or more this would really make a difference (and even with just one consumer, you can already remove two Promise.resolve() invocations that the transpiler adds for you.

File: packages/firestore/src/core/sync_engine.ts (hunk 8 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/core/sync_engine.ts]: You can invoke `yarn build`  in "packages/firestore". It will spit out a couple different files into "packages/firestore/dist"

In packages/firestore/dist/index.esm.js, your current code looks as such:

```
callbacks = this.pendingWritesCallbacks.get(highestBatchId) || [];
callbacks.push(callback);
this.pendingWritesCallbacks.set(highestBatchId, callbacks);
```

It looks like we have to specify different compiler settings to make my suggestion work (--downlevelIterators)/ Since it is just an optional suggestion, let's keep it as is.

File: packages/firestore/src/api/database.ts [hunk 22]
```diff
@@ -478,6 +478,25 @@ export class Firestore implements firestore.FirebaseFirestore, FirebaseService {
     return this._firestoreClient!.clientShutdown;
   }
 
+  /**
+   * Waits until all currently pending writes for the active user have been acknowledged by the
+   * backend.
+   *
+   * The returned Promise resolves immediately if there are no outstanding writes. Otherwise, the
+   * Task waits for all previously issued writes (including those written in a previous app
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts]: s/Task/Promise/

File: packages/firestore/src/api/database.ts [hunk 23]
```diff
@@ -478,6 +478,25 @@ export class Firestore implements firestore.FirebaseFirestore, FirebaseService {
...
+   * session), but it does not wait for writes that were added after the method is called. If you
+   * wish to wait for additional writes, you have to call `waitForPendingWrites()` again.
+   *
+   * Any outstanding `waitForPendingWrites()` Promises are rejected during user changes.
+   *
+   * @return A `Promise` which resolves when all currently pending writes have been
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts]: Nit: remove backticks.

File: packages/firestore/src/core/firestore_client.ts [hunk 24]
```diff
@@ -522,6 +522,21 @@ export class FirestoreClient {
     });
   }
 
+  /**
+   * Returns a Promise that resolves when when all writes that were pending at the time when this
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts]: s/when when/when/
s/at the time when/at the time/

File: packages/firestore/src/core/firestore_client.ts [hunk 25]
```diff
@@ -522,6 +522,21 @@ export class FirestoreClient {
...
+   * method is called received server acknowledgement. An acknowledgement can be either acceptance
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts]: s/is/was/

File: packages/firestore/src/core/firestore_client.ts [hunk 26]
```diff
@@ -522,6 +522,21 @@ export class FirestoreClient {
...
+   * or rejections.
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/firestore_client.ts]: s/rejections/rejection/

File: packages/firestore/src/core/sync_engine.ts [hunk 27]
```diff
@@ -450,6 +452,7 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
       (this.isPrimary && source === OnlineStateSource.RemoteStore) ||
       (!this.isPrimary && source === OnlineStateSource.SharedClientState)
     ) {
+      this.assertSubscribed('applyOnlineStateChange()');
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: Is this intended?

File: packages/firestore/src/core/sync_engine.ts [hunk 28]
```diff
@@ -602,6 +609,59 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     }
   }
 
+  /**
+   * Registers a user callback with all the pending mutations at the moment of calling.
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/sync_engine.ts]: Can you update this to:

Registers a user callback that resolves when all pending mutations at the moment of calling are acknowledged .

You can then remove the second sentence.

File: packages/firestore/src/core/sync_engine.ts (hunk 10 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/core/sync_engine.ts]: I learned something today :)

File: packages/firestore/test/integration/api/database.test.ts [hunk 29]
```diff
@@ -1120,4 +1123,49 @@ apiDescribe('Database', (persistence: boolean) => {
...
+  });
+
+  it('waiting for pending writes resolves immediately when offline and no pending writes',
+    async () => {
+      await withTestDoc(persistence, async docRef => {
+        const firestore = docRef.firestore;
+        // Prevent pending writes receiving acknowledgement.
+        await firestore.disableNetwork();
+
+      // `awaitsPendingWrites` is created when there is no pending writes, it will resolve
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: This looks like it needs reformatting.

File: packages/firestore/test/integration/api/database.test.ts [hunk 30]
```diff
@@ -1120,4 +1123,49 @@ apiDescribe('Database', (persistence: boolean) => {
...
+      // immediately even if we are offline.
+        const awaitPendingWrites = waitForPendingWrites(firestore);
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/database.test.ts]: Suggest to remove the redundant variable here.

**Comment by schmidt-sebastian** [comment]
> I do want you to make `registerPendingWritesCallback` non-async though :)

Looks like this is not possible as `getHighestUnacknowledgedBatchId` returns a Promise. Thanks for pointing this out.

File: packages/firestore/src/api/database.ts (hunk 22 above)
Comment (wu-hui) [reply packages/firestore/src/api/database.ts]: Done.

File: packages/firestore/src/api/database.ts (hunk 23 above)
Comment (wu-hui) [reply packages/firestore/src/api/database.ts]: Done.

File: packages/firestore/src/core/firestore_client.ts (hunk 24 above)
Comment (wu-hui) [reply packages/firestore/src/core/firestore_client.ts]: Done.

File: packages/firestore/src/core/firestore_client.ts (hunk 25 above)
Comment (wu-hui) [reply packages/firestore/src/core/firestore_client.ts]: Done.

File: packages/firestore/src/core/firestore_client.ts (hunk 26 above)
Comment (wu-hui) [reply packages/firestore/src/core/firestore_client.ts]: Done.

File: packages/firestore/src/core/sync_engine.ts (hunk 27 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts]: Yes. This was missing but should be present. Otherwise `this.syncEngineListener!.*` might hit errors.

File: packages/firestore/src/core/sync_engine.ts (hunk 28 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts]: Done.

File: packages/firestore/src/core/sync_engine.ts (hunk 7 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts]: `this.localStore.getHighestUnacknowledgedBatchId()` returns a promise, which demands it to be async.

File: packages/firestore/src/core/sync_engine.ts (hunk 10 above)
Comment (wu-hui) [reply packages/firestore/src/core/sync_engine.ts]: :)

File: packages/firestore/test/integration/api/database.test.ts (hunk 29 above)
Comment (wu-hui) [reply packages/firestore/test/integration/api/database.test.ts]: Done.

File: packages/firestore/test/integration/api/database.test.ts (hunk 30 above)
Comment (wu-hui) [reply packages/firestore/test/integration/api/database.test.ts]: Done.
